# [CptS 215 Data Analytics Systems and Algorithms](https://piazza.com/wsu/fall2017/cpts215/home)
[Washington State University](https://wsu.edu)

[Gina Sprint](http://eecs.wsu.edu/~gsprint/)
## MA4 Queue Analysis (50 pts)
<mark>Due:</mark>

### Learner Objectives
At the conclusion of this micro assignment, participants should be able to:
* Analyze stack and queue data structures for efficiency
* Implement a queue using two stacks
* Compare/contrast different implementations of a queue ADT

### Prerequisites
Before starting this micro assignment, participants should be able to:
* Write object oriented Python code
* Write Markdown and code cells in Jupyter Notebook
* Understand the stack ADT and implement a stack
* Understand the queue ADT and implement a queue

### Acknowledgments
Content used in this assignment is based upon information in the following sources:
* Carl Kingsford's [Basic Data Structures](https://www.engage-csedu.org/find-resources/problem-set-1-basic-data-structures) problem set.

## Overview and Requirements
For this micro assignment, you are going to download this Jupyter Notebook and answer the following questions. Your answer for a problem should be in a cell *immediately* after the problem cell. *Do not modify the problem cell.*

We are going to explore the efficiency of two different queue implementations. This micro assignment includes conceptional questions and programming.

### Conceptual Questions
Suppose you are given a (strange) computer that can only perform the following instructions (in addition to if and while):
* `S = create_stack()` create stack makes a new stack `S`
* `i = S.pop()` removes the top item from stack `S` and places it in variable `i`
* `S.push(i)` makes item `i` the top item in stack `S`

Solve the following problems and *justify* your answers:
1. (10 pts) Show how you can use these operations to implement a queue (operations `Q = create_queue()`, `enqueue(i)`, `i = dequeue()`)
    * A picture might help to explain your answer
    * Hint: take a look at the following image:
<img src="http://www.algoqueue.com/algoqueue/members/get_uploaded_image.load/149" width="500">
(image from [http://www.algoqueue.com/algoqueue/members/get_uploaded_image.load/149](http://www.algoqueue.com/algoqueue/members/get_uploaded_image.load/149))
1. (5 pts) What's the worst case running time of your dequeue implementation?
1. (5 pts) Over a series of `n` enqueues followed by `n` dequeues, how many `pop()` operations does your implementation perform?

### Implementation Questions
Write a program that implements a queue using a standard list implementation (see [M&R 3.12](http://interactivepython.org/runestone/static/pythonds/BasicDS/ImplementingaQueueinPython.html)) and a queue using your solution to conceptual question #1. For the latter, you must implement a stack using a standard list implementation (see [M&R 3.5](http://interactivepython.org/runestone/static/pythonds/BasicDS/ImplementingaStackinPython.html)).

Generate a sequence of `enqueue`s and `dequeue`s to test your two queue implementations. To generate the test sequence, randomly `enqueue` and `dequeue` strings from [words.txt](https://raw.githubusercontent.com/gsprint23/cpts215/master/microassignments/files/words.txt), a file containing all 118,309 valid crossword puzzle words, one on each line. Evaluate the differences between the two implementations by performing the following:
1. Using [`timeit()`](https://docs.python.org/3/library/timeit.html), compare the running time for each queue implementation operating on your test sequence. Vary the size of your test sequence.
    * Note: Make sure you are using the same test sequence for each implementation! Also, remove all frivolous code from your implementations (e.g. `print()` statements), as these can affect the timing!
1. Write code to test your answer to conceptual question #2. Write up your observations.
1. Write code to test your answer to conceptual question #3. Write up your observations.

# Conceptual Questions Answer/Reponses
(1) We can use two stacks for implementing a queue.
    
```
stack_1 = create_stack()
stack_2 = create_stack()
```

When we initially push data into `stack_1`, the first item will be at the bottom while the last item is on the top (LIFO order). By using a second stack, we can reverse this order again by push()'ing all of the pop() operations from `stack_1` to `stack_2`. By doing this, we rearrange the data in `stack_2` in the order of which they were push() in `stack_1`, so when you pop() `stack_2`, you get items in a FIFO-order. That is: the first item which was push() in `stack_1` is the first item that is pop() from `stack_2`, the second item which was push() in `stack_1` is the first item that is pop() from `stack_2`, ... etc.


```
stack_1.push(data)
... # push all data into stack_1
stack_2.push(stack_1.pop())
... # then push all the data from stack_1 into stack_2 using the pop() data from stack_1
stack_2.pop() # now that all the data is in stack_2, pop()'ing stack_2 should return all the items in a nature of the queue (FIFO)
```
(2) Worst case of dequeuing with two stacks should be O(n). This is because you have to transfer all of the data into the first stack with push(), then pop() them from the first stack as you push() the pop()'ed item from the first stack to the second stack. Then to dequeue, you simply pop() `stack_2`. All these operations are linear in nature, and so their total time complexity should amount to something resembling a linear trend O(n).

(3) It would be 2n. The first set of pop() is for `stack_1` to transfer data to the second stack, which requires n pop(), then the second set of pop() is for `stack_2` which requires another n-pop(). Add them together $n+n$ and we get $2n$ total pop() for the implementation.

In [186]:
class Queue:
    def __init__(self):
        self.items = []

    def isEmpty(self):
        return self.items == []

    def enqueue(self, item):
        self.items.insert(0,item)

    def dequeue(self):
        return self.items.pop()

    def size(self):
        return len(self.items)
    
class Stack:
    def __init__(self):
        self.items = []

    def isEmpty(self):
        return self.items == []

    def push(self, item):
        self.items.append(item)

    def pop(self):
        return self.items.pop()

    def peek(self):
        return self.items[len(self.items)-1]

    def size(self):
        return len(self.items)

def setup_list():
    global wordlist
    wordlist = open('words.txt', 'r')


def run_queue(data_size):
    queue = Queue()
    
    # Populate the queue
    j = 0
    for i in wordlist:
        if(j >= data_size): break
        queue.enqueue(i)
        j += 1
        
    # Dequeue
    for i in range(0,data_size):
        queue.dequeue()
    
def run_stack(data_size):
    stack_1 = Stack()
    stack_2 = Stack()
    
    # Populate initial stack with data
    j = 0
    for i in wordlist:
        if(j >= data_size): break
        stack_1.push(i)
        j += 1
    # Move data from initial stack to second stack
    for i in range(0,data_size):
        stack_2.push(stack_1.pop())
                   
    # Dequeue the second stack
    for i in range(0,data_size):
        stack_2.pop()
                   
if __name__ == '__main__':
    import timeit
    
    time_stack_1 = timeit.timeit("run_stack(100), setup_list()", setup="from __main__ import setup_list, run_stack", number=1)
    time_stack_2 = timeit.timeit("run_stack(1000), setup_list()", setup="from __main__ import setup_list, run_stack", number=1)
    time_stack_3 = timeit.timeit("run_stack(100000), setup_list()", setup="from __main__ import setup_list, run_stack", number=1)
                   
    time_queue_1 = timeit.timeit("run_queue(100), setup_list()", setup="from __main__ import setup_list, run_queue", number=1)
    time_queue_2 = timeit.timeit("run_queue(1000), setup_list()", setup="from __main__ import setup_list, run_queue", number=1)
    time_queue_3 = timeit.timeit("run_queue(100000), setup_list()", setup="from __main__ import setup_list, run_queue", number=1)
    
    print("For data_size = 100, 1000, 100000, it took the stack implementation of the queue: %0.9f, %0.9f, %0.9f seconds respectively to run." %(time_stack_1,time_stack_2,time_stack_3))
    print("For data_size = 100, 1000, 100000, it took the queue implementation: %0.9f, %0.9f, %0.9f seconds respectively to run." %(time_queue_1,time_queue_2,time_queue_3))

For data_size = 100, 1000, 100000, it took the stack implementation of the queue: 0.000592250, 0.001280191, 0.110734523 seconds respectively to run.
For data_size = 100, 1000, 100000, it took the queue implementation: 0.000243300, 0.001185105, 2.924262585 seconds respectively to run.


In [185]:
# Answering implementation 2

setup_list()
time_stack_1 = timeit.timeit("run_stack(1000), setup_list()", setup="from __main__ import setup_list, run_stack", number=1)
time_stack_2 = timeit.timeit("run_stack(10000), setup_list()", setup="from __main__ import setup_list, run_stack", number=1)
time_stack_3 = timeit.timeit("run_stack(100000), setup_list()", setup="from __main__ import setup_list, run_stack", number=1)

print("n=1000, time=%0.9f" %time_stack_1)
print("n=10000, time=%0.9f" %time_stack_2)
print("n=100000, time=%0.9f" %time_stack_3)

n=1000, time=0.002320100
n=10000, time=0.016928340
n=100000, time=0.113570201


### Answer to implementation 2
As we see, the time in relation to the size of data appears to be linear. For 1000 items, we have 0.002320100 seconds. For 10,000 items, we have 0.016928340. And lastly for 100,000 items, we have 0.113570201. The time increases by what appears to be a factor of ~7. Thus the time complexity is said to be linear, O(n).

In [ ]:
def pop_in_stack(data_size):
    stack_1 = Stack()
    stack_2 = Stack()
    
    # Populate initial stack with data
    j = 0
    for i in wordlist:
        if(j >= data_size): break
        stack_1.push(i)
        j += 1
    
    sum_of_pop = 0
    # Move data from initial stack to second stack
    for i in range(0,data_size):
        stack_2.push(stack_1.pop())
        sum_of_pop += 1
        
    # Dequeue the second stack
    for i in range(0,data_size):
        stack_2.pop()
        sum_of_pop += 1
    
    # in other words, data_size + data_size = sum_of_pop
    return sum_of_pop

# Answering implementation question 3
n = 10
retval = pop_in_stack(n)
print(retval)

n = 100
retval = pop_in_stack(n)
print(retval)

n = 200
retval = pop_in_stack(n)
print(retval)

### Answer to implementation 3.
As we see in the body of pop_in_stack() (a modified run_stack()), the pop operations occur twice in for loops that runs from 0 to data_size (which is also n-data). So, we could think of sum_of_pop as just a summation of how many times the loops iterate which is simply `data_size + data_size = 2*data_size`.

## Bonus (5 pts)
Perform additional analysis and comparisons of your two queue implementations above to the following:
1. Linked lists implementations of the stacks and queue (you must write the implementation yourself)
1. Python's `deque` container from `collections`

Include `timeit()` results and a write up of your observations.

## Submitting Assignments
1.	Use the Blackboard tool https://learn.wsu.edu to submit your assignment. You will submit your code to the corresponding programming assignment under the "Content" tab. You must upload your solutions as `<your last name>_ma4.zip` by the due date and time.
2.	Your .zip file should contain your .ipynb file and a .html file representing your Notebook as a webpage (File->Download as->HTML). Also include [words.txt](https://raw.githubusercontent.com/gsprint23/cpts215/master/microassignments/files/words.txt) in your .zip file.

## Grading Guidelines
This assignment is worth 50 points + 5 points bonus. Your assignment will be evaluated based on a successful compilation and adherence to the program requirements. We will grade according to the following criteria:
* 20 pts for answering the conception questions
* 5 pts for correct implementation of the list-based queue
* 10 pts for correct implementation of the stack-based queue
* 5 pts for `timeit()` results
* 5 pts for observation write ups
* 5 pts for for adherence to proper programming style and comments established for the class